### PR TITLE
feat(eif): sign, repack, and resign EIFs via dedicated [eif] keyset

### PR DESCRIPTION
**Pull request by jmt-lab**
Extend eif-builder with EifSectionSignature (0x04) covering PCR0 via COSE_Sign1 ECDSA (ES256/ES384). rpm2eif reads a dedicated `[eif]` section in Infra.toml and picks local (aws-lc-rs) or KMS signing from its shape; absent section builds unsigned. Replaces the earlier sbkeys autodetect so the Nitro-hypervisor trust root is decoupled from the UEFI Secure Boot code-signing key (distinct algorithms, rotation cadence, and trust domain). Signing is orthogonal to UEFI Secure Boot.

Wire format matches aws-nitro-enclaves-image-format: CBOR Array(1) of Map { signing_certificate: PEM, signature: COSE_Sign1 } with a PcrInfo payload for PCR0 = SHA-384(48 zero bytes || SHA-384(kernel || cmdline || ramdisks)), so nitro-cli describe-eif populates PCR8.

Adds two repack-variant capabilities:
* In-tree EIF repack (eif2eif): rewrites CA bundle / root.json in the EIF rootfs, rebuilds erofs + dm-verity, and produces a freshly signed .eif with a cmdline pinned to the new verity root hash.
* Guest EIF resign during host repack: img2img walks guest-images install paths and resigns every *.eif in place via a new `eif-builder resign` subcommand (byte-preserves kernel/cmdline/ramdisk, rewrites only the signature section + header CRC).

Highlights:
* new eif-builder/src/signer.rs (Signer trait, LocalSigner, KmsSigner)
* CLI: --signing-cert + --signing-key (local) or --kms-key-id (KMS), enforced via clap ArgGroup; new `resign` subcommand
* typed KMS errors (KmsAccessDenied, KmsKeyDisabled, KmsInvalidKeyState)
* ECDSA DER->raw validates SEQUENCE length and rejects trailing bytes
* shared eif-sign-helper sourced by rpm2eif, eif2eif, and img2img; hard-fails on mounted cert without a signing backend (opt-out: EIF_ALLOW_UNSIGNED_WITH_CERT=1)

New deps: aws-lc-rs, coset, ciborium, pem, x509-cert.

Tests: PCR0 spec conformance, COSE payload shape, end-to-end signature verification, DER parser edge cases, resign round-trips, and the eif-sign-helper discovery matrix.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: tests/integration-tests/src/variant_build.rs [hunk 1]
```diff
@@ -138,3 +138,217 @@ async fn test_twoliter_repack_variant() {
...
+/// This exercises the `img2img --guest-images=...` path added by this change:
+/// the host repack must walk each declared guest install path under the
+/// extracted rootfs, resign each `*.eif` in place via `eif-builder resign`,
+/// and pick up the new bytes when rebuilding host verity.
+#[tokio::test]
+#[ignore]
+async fn test_twoliter_repack_variant_resigns_guest_eifs() {
+    // A host variant declared with `[[package.metadata.build-variant.guest-images]]`
+    // pointing at an EIF guest. Adjust to whichever upstream host+guest pair
+    // currently ships in bottlerocket-os.
```
Comment (ginglis13) [inline tests/integration-tests/src/variant_build.rs:240]:
```suggestion
    // pointing at an EIF guest.
```

File: tests/integration-tests/src/variant_build.rs [hunk 2]
```diff
@@ -138,3 +138,217 @@ async fn test_twoliter_repack_variant() {
...
+    assert!(kernel_count >= 1, "no -kernel under {}", out_dir.display());
+}
+
+/// Repack a host variant that embeds a guest EIF, and assert the guest EIF's
+/// signature section changed (bytewise) while the guest kernel bytes did not.
+///
+/// This exercises the `img2img --guest-images=...` path added by this change:
+/// the host repack must walk each declared guest install path under the
+/// extracted rootfs, resign each `*.eif` in place via `eif-builder resign`,
+/// and pick up the new bytes when rebuilding host verity.
```
Comment (ginglis13) [inline tests/integration-tests/src/variant_build.rs:234]:
```suggestion
/// This exercises the `img2img --guest-images=...` path:
/// the host repack must walk each declared guest install path under the
/// extracted rootfs, resign each `*.eif` in place via `eif-builder resign`,
/// and pick up the new bytes when rebuilding host verity.
```

File: tests/integration-tests/src/variant_build.rs [hunk 3]
```diff
@@ -138,3 +138,217 @@ async fn test_twoliter_repack_variant() {
...
+/// bottlerocket-os repo at the time this test is run. If the upstream
+/// stops shipping it, replace with any current `image-format = "eif"`
+/// variant (or bring one into the local test project).
+#[tokio::test]
+#[ignore]
+async fn test_twoliter_repack_variant_eif() {
+    // Placeholder variant name; the upstream repo carries `aws-nitro-eks-2`
+    // and similar EIF variants under `variants/`. Any variant whose
+    // `image-format = "eif"` will exercise the same path.
+    const EIF_VARIANT: &str = "aws-nitro-eks-2";
```
Comment (ginglis13) [inline tests/integration-tests/src/variant_build.rs:155]: ?

File: tests/integration-tests/src/variant_build.rs [hunk 4]
```diff
@@ -138,3 +138,217 @@ async fn test_twoliter_repack_variant() {
...
     );
 }
+
+/// Repack an EIF-format variant end-to-end. Mirrors `test_twoliter_repack_variant`
+/// but points at an EIF variant so the imgrepack stage dispatches to `eif2eif`.
+///
+/// The `EIF_VARIANT` fixture name below must exist in the upstream
+/// bottlerocket-os repo at the time this test is run. If the upstream
+/// stops shipping it, replace with any current `image-format = "eif"`
+/// variant (or bring one into the local test project).
```
Comment (ginglis13) [inline tests/integration-tests/src/variant_build.rs:148]:
```suggestion
```

File: tests/integration-tests/src/variant_build.rs [hunk 5]
```diff
@@ -138,3 +138,217 @@ async fn test_twoliter_repack_variant() {
...
+    let output = twoliter_make(bob_src.path(), "repack-variant", EIF_VARIANT, arch);
+    assert!(
+        output.status.success(),
+        "twoliter make repack-variant (eif) failed: {}",
+        String::from_utf8_lossy(&output.stderr)
+    );
+
+    // The repacked EIF must exist under the expected build dir. Assert on
+    // real regular files, not on directory presence, so a regression that
+    // makes `eif2eif` silently produce no artifacts fails this test.
```
Comment (ginglis13) [inline tests/integration-tests/src/variant_build.rs:197]:
```suggestion
```

File: tests/integration-tests/src/variant_build.rs [hunk 6]
```diff
@@ -138,3 +138,217 @@ async fn test_twoliter_repack_variant() {
...
+        String::from_utf8_lossy(&output.stderr)
+    );
+
+    // Snapshot the SHA-256 of the host's `.img.lz4` output before repack.
+    // The host embeds guest EIFs inside its ROOT-A; when img2img resigns
+    // any of them, the outer compressed image bytes will differ. This is
+    // a weaker signal than extracting the guest EIF and comparing its
+    // signature section directly (which would require unpacking the host
+    // rootfs — non-trivial), but it does distinguish "repack ran and
+    // touched the rootfs" from "repack was a no-op".
```
Comment (ginglis13) [inline tests/integration-tests/src/variant_build.rs:280]: Reading the code and assertion method is fairly self explanatory here
```suggestion
```

File: tools/buildsys/src/builder/error.rs [hunk 7]
```diff
@@ -84,6 +84,42 @@ pub(crate) enum Error {
...
+    InfraConfigLoad {
+        path: PathBuf,
+        #[snafu(source(from(pubsys_config::Error, Box::new)))]
+        source: Box<pubsys_config::Error>,
+    },
+
+    #[snafu(display(
+        "EIF signing cert '{}' does not exist or is empty (from Infra.toml [eif].signing_cert). \
+         Note: an empty file is rejected because the in-container helper's `[[ -s ]]` gate would \
+         silently treat it as \"no profile\" and produce an unsigned EIF.",
```
Comment (ginglis13) [inline tools/buildsys/src/builder/error.rs:97]: Not sure this verbosity is needed
```suggestion
```

File: tools/buildsys/src/builder.rs [hunk 8]
```diff
@@ -213,6 +214,27 @@ impl crate::builder::PackageBuildArgs {
...
 
+/// Format the variant-declared EIF PCIE flags for the Docker build-arg.
+///
+/// Returns the value as bare lowercase hex (no `0x` prefix); `eif-builder`'s
+/// `--pcie-flags` parser accepts the prefixed and unprefixed forms
+/// interchangeably, and stripping it keeps the ARG stable for docker layer
+/// caching regardless of how the variant author wrote the literal in
+/// `Cargo.toml`. An unset field becomes the empty string, which the
+/// Dockerfile treats as "omit `--pcie-flags` and let `eif-builder` apply its
+/// `DEFAULT_PCIE_FLAGS`".
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:225]:
```suggestion
```

File: tools/buildsys/src/builder.rs [hunk 9]
```diff
@@ -213,6 +214,27 @@ impl crate::builder::PackageBuildArgs {
...
+fn format_eif_pcie_flags(flags: Option<u16>) -> String {
+    flags.map(|v| format!("{v:x}")).unwrap_or_default()
+}
+
+/// Serialize the variant-authored EIF kernel format for the `EIF_KERNEL_FORMAT`
+/// build-arg. When unset we forward the empty string; the Dockerfile passes
+/// that through unchanged and `rpm2eif` treats an empty value as "apply my
+/// built-in default" (bzImage on x86_64). Non-EIF variants ignore this.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:233]:
```suggestion
/// build-arg.
```

File: tools/buildsys/src/builder.rs [hunk 10]
```diff
@@ -223,6 +245,19 @@ struct VariantBuildArgs {
     image_features: HashSet<ImageFeature>,
     image_format: String,
     kernel_parameters: String,
+    /// EIF header PCIE flags override from `[package.metadata.build-variant]
+    /// eif-pcie-flags`. Forwarded to `rpm2eif` as the `EIF_PCIE_FLAGS` build-arg
+    /// (hex, no `0x` prefix). Empty string when the variant does not set the
+    /// field, in which case `rpm2eif` omits `--pcie-flags` and `eif-builder`
+    /// applies its own `DEFAULT_PCIE_FLAGS`. Non-EIF variants ignore this.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:252]:
```suggestion
    /// EIF header PCIE flags override from `[package.metadata.build-variant]
    /// eif-pcie-flags`.
```

File: tools/buildsys/src/builder.rs [hunk 11]
```diff
@@ -223,6 +245,19 @@ struct VariantBuildArgs {
...
+    eif_pcie_flags: String,
+    /// x86_64 EIF kernel format override from `[package.metadata.build-variant]
+    /// eif-kernel-format`. Forwarded to `rpm2eif` as the `EIF_KERNEL_FORMAT`
+    /// build-arg (`"bzimage"` or `"vmlinux"`). Empty string when unset, in
+    /// which case `rpm2eif` applies its own default (bzImage — the sidecar
+    /// enclave loader's boot protocol). Ignored on aarch64 and by non-EIF
+    /// variants.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:259]:
```suggestion
    /// x86_64 EIF kernel format override from `[package.metadata.build-variant]
    /// eif-kernel-format`.
```

File: tools/buildsys/src/builder.rs [hunk 12]
```diff
@@ -314,6 +351,19 @@ struct RepackVariantBuildArgs {
     data_image_size_gib: String,
     image_features: HashSet<ImageFeature>,
     image_format: String,
+    /// Space-joined variant-declared kernel parameters. Required at repack time
+    /// for `image-format = "eif"` because `eif2eif` reconstructs the kernel
+    /// command line from scratch (new dm-verity root hash ⇒ new PCR0 ⇒ full
+    /// rebuild), and any variant-supplied parameters must survive. Mirrors
+    /// `VariantBuildArgs::kernel_parameters`. Empty when the variant declares
+    /// none; harmless for the non-EIF (img2img) path, which does not touch
+    /// the cmdline.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:360]:
```suggestion
    /// Space-joined variant-declared kernel parameters. Required at repack time
    /// for `image-format = "eif"`
```

File: tools/buildsys/src/builder.rs [hunk 13]
```diff
@@ -314,6 +351,19 @@ struct RepackVariantBuildArgs {
...
+    kernel_parameters: String,
+    /// Mirrors `VariantBuildArgs::eif_pcie_flags`. Required at repack time for
+    /// `image-format = "eif"` because `eif2eif` rebuilds the EIF header from
+    /// scratch and must reapply the variant-authored value. Empty when unset;
+    /// harmless on the non-EIF (img2img) path, which does not touch the header.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:365]:
```suggestion
    /// Mirrors `VariantBuildArgs::eif_pcie_flags`. Required at repack time for
    /// `image-format = "eif"`
```

File: tools/buildsys/src/builder.rs [hunk 14]
```diff
@@ -596,18 +664,22 @@ impl DockerBuild {
...
 
     /// Create a new `DockerBuild` that can repackage a variant image.
-    pub(crate) fn repack_variant(args: RepackVariantArgs, manifest: &Manifest) -> Result<Self> {
+    ///
+    /// `guest_images` is the resolved list of guest variants declared by the
+    /// host under `[package.metadata.build-variant.guest-images]`. For a
+    /// non-EIF host, these are the guest install paths whose `*.eif` files
+    /// `img2img` must resign before rebuilding host verity. Empty for a
+    /// variant that declares no guests (or for an EIF variant, which has
+    /// no rootfs to walk).
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:673]:
```suggestion
```

File: tools/buildsys/src/builder.rs [hunk 15]
```diff
@@ -596,18 +664,22 @@ impl DockerBuild {
...
-        // fast here so users get a clear error instead of a cryptic failure
-        // deep inside the Dockerfile build.
-        if matches!(manifest.info().image_format(), Some(ImageFormat::Eif)) {
-            return error::EifRepackUnsupportedSnafu.fail();
-        }
-
-        // Repack has already rejected EIF above, so `image_format` is always
-        // non-EIF here; pass it through anyway for defense-in-depth.
+        // Repack now supports EIF: for image-format="eif" the imgrepack
+        // Dockerfile stage dispatches to `eif2eif` instead of `img2img`.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:682]:
```suggestion
        // When image-format="eif" the imgrepack
        // Dockerfile stage dispatches to `eif2eif`, otherwise `img2img`.
```

File: tools/buildsys/src/builder.rs [hunk 16]
```diff
@@ -958,9 +1049,198 @@ fn secrets_args() -> Result<Vec<String>> {
...
+///   `eif-kms-key-id.env`  – KMS key id/ARN/alias, sourced from the buildsys
+///                           process env (KMS backend only)
+///
+/// A missing Infra.toml, a missing `[eif]` section, or an unset
+/// `PUBLISH_INFRA_CONFIG_PATH` is *not* an error: EIF variants without a
+/// configured signing profile fall back to producing unsigned artifacts (the
+/// helper in `eif-sign-helper` interprets missing mounts as "unsigned").
+/// This matches the pre-existing "no code-sign.crt ⇒ unsigned" contract that
+/// local dev, kits, and non-nitro variants depend on.
+fn eif_signing_args() -> Result<Vec<String>> {
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:1080]: please revise this method with more concise comments and docstring.

File: tools/buildsys/src/builder.rs [hunk 17]
```diff
@@ -958,9 +1049,198 @@ fn secrets_args() -> Result<Vec<String>> {
...
+        // secret" as "no profile."
+        detect_stale_infra_lock(&infra_path)?;
+        return Ok(Vec::new());
+    };
+
+    // Existence + non-empty checks up-front so failure surfaces here
+    // (buildsys) rather than mid-build inside the container, where the
+    // operator sees a bash error with far less context. Docker itself will
+    // fail if a secret source is missing, but its error message doesn't
+    // identify Infra.toml as the source of the path.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:1114]: case in point for the verbosity, we shouldn't be discussing style philosophy in comments IMO

File: tools/buildsys/src/builder.rs [hunk 18]
```diff
@@ -958,9 +1049,198 @@ fn secrets_args() -> Result<Vec<String>> {
...
+/// Mirrors the semantics of bash `[[ -s <path> ]]`, which is the exact gate
+/// used by `eif-sign-helper` to decide whether a signing profile is present.
+/// Keeping the two sides in lockstep is what prevents the "buildsys says
+/// present, helper says empty ⇒ silent unsigned" split-brain failure mode.
+///
+/// A missing file, an unreadable file, a directory, or a 0-byte regular
+/// file all return `false`. We deliberately do not propagate `metadata()`
+/// errors: the caller only needs a yes/no answer, and any I/O error at
+/// this stage will re-surface with more context when Docker itself tries
+/// to mount the path as a build secret.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:1184]: same here - could be simplified to just
```
/// Return boolean result of if file at `path` is empty
```

File: tools/buildsys/src/main.rs [hunk 19]
```diff
@@ -306,11 +306,20 @@ fn repack_variant(args: RepackVariantArgs) -> Result<()> {
     check_arch_support(manifest.info(), args.common.arch);
     validate_standalone_image_or_warn(manifest.info())?;
 
+    // Resolve declared guest-images. On a repack of a host variant, the
+    // imgrepack stage passes these install paths through to `img2img` so
+    // it can find and resign every `*.eif` under each install path before
+    // rebuilding host verity. For an EIF-format variant (which has no
+    // guests inside its own rootfs) the list is empty.
```
Comment (ginglis13) [inline tools/buildsys/src/main.rs:313]: seems unnecessary to embed this comment here; docs on structure of embedded guest images, and how that is handled for repack / eif2eif, have use covered.

File: tools/buildsys/src/manifest.rs [hunk 20]
```diff
@@ -629,6 +673,30 @@ impl ManifestInfo {
...
     }
 
+    /// Convenience method to return the EIF header PCIE flags override for
+    /// this variant. Only meaningful when `image-format = "eif"`; ignored
+    /// otherwise. Returned as-is (a `u16`); the value written into the EIF
+    /// header is `arch_flags | eif_pcie_flags`, matching what
+    /// `eif-builder --pcie-flags <hex>` does internally.
+    ///
+    /// See the `eif_pcie_flags` field on [`BuildVariant`] for the
+    /// header==launch-flags contract this knob exists to author.
```
Comment (ginglis13) [inline tools/buildsys/src/manifest.rs:683]:
```suggestion
    /// Convenience method to return the EIF header PCIE flags override for
    /// this variant. Only meaningful when `image-format = "eif"`; ignored
    /// otherwise.
```

File: tools/buildsys/src/manifest.rs [hunk 21]
```diff
@@ -629,6 +673,30 @@ impl ManifestInfo {
...
+    }
+
+    /// Convenience method to return the x86_64 EIF kernel format override
+    /// for this variant. Only meaningful when `image-format = "eif"` on
+    /// x86_64; ignored otherwise. Returns `None` when unset, in which case
+    /// `rpm2eif` applies its built-in default (bzImage — the sidecar
+    /// enclave loader's boot protocol).
+    ///
+    /// See the `eif_kernel_format` field on [`BuildVariant`] for why this
+    /// stays selectable.
```
Comment (ginglis13) [inline tools/buildsys/src/manifest.rs:695]:
```suggestion
    /// Convenience method to return the x86_64 EIF kernel format override
    /// for this variant. Only meaningful when `image-format = "eif"` on
    /// x86_64; ignored otherwise.
```

**Review by ginglis13 — COMMENTED** [review]
just an initial pass, mostly nitpicking overly verbose comments

File: tests/integration-tests/src/variant_build.rs (hunk 3 above)
Comment (jmt-lab) [reply tests/integration-tests/src/variant_build.rs:154]: We should change this and remove mention of upstream repo which isn't factual and doesn't mean anything here

File: tools/buildsys/src/builder.rs [hunk 22]
```diff
@@ -958,9 +1025,125 @@ fn secrets_args() -> Result<Vec<String>> {
...
+/// Detection is intentionally coarse: any parseable Infra.toml with an
+/// `[eif]` section, next to an `Infra.lock` that was actually loaded, is
+/// treated as stale. We don't try to diff the two — if the lock is
+/// current, the caller already unwrapped `Some(eif)` and never reached
+/// this function.
+///
+/// If Infra.toml is malformed, missing, or the paths don't line up, we
+/// stay quiet: `from_path_or_lock` already validated whichever file it
+/// actually loaded, and the "no eif" outcome is the correct one for the
+/// no-Infra.toml case.
```
Comment (ginglis13) [inline tools/buildsys/src/builder.rs:1114]:
```suggestion
/// Ensure Infra.lock file is accurate and up-to-date
```

File: tools/buildsys/src/manifest.rs [hunk 23]
```diff
@@ -937,6 +995,58 @@ pub enum SensitivityType {
...
+/// written in hex on the `eif-builder --pcie-flags` CLI (whose parser is
+/// hex-only, prefix optional) and in the aws-nitro-enclaves-image-format
+/// header definitions. A user copy-pasting a hex value like `340` from those
+/// contexts into `eif-pcie-flags = "340"` almost certainly means `0x340`
+/// (bits 6, 8, 9), not `340` decimal (`0x154`, bits 2, 4, 6, 8). Treating
+/// the unprefixed string as decimal would produce a header/launch-flags
+/// mismatch that only surfaces at enclave launch, which is the exact kind
+/// of silent misconfiguration this knob exists to prevent. Forcing an
+/// explicit base makes the author's intent unambiguous at parse time:
+/// integer form is unambiguously decimal, string form is unambiguously hex.
```
Comment (ginglis13) [inline tools/buildsys/src/manifest.rs:1012]: This is already covered by the crate level docs

```suggestion
/// silently interpreted.
```

File: tools/buildsys/src/manifest.rs [hunk 24]
```diff
@@ -964,6 +1074,54 @@ pub struct BuildVariant {
...
+    /// flags from the header directly (host echoes, hypervisor enforces), the
+    /// coupling goes away — but the knob is still how the header value gets
+    /// authored per-variant.
+    ///
+    /// Example:
+    /// ```ignore
+    /// [package.metadata.build-variant]
+    /// image-format = "eif"
+    /// eif-pcie-flags = 0x340
+    /// ```
```
Comment (ginglis13) [inline tools/buildsys/src/manifest.rs:1105]: Already covered in crate documentation
```suggestion
```

File: tools/buildsys/src/manifest.rs [hunk 25]
```diff
@@ -964,6 +1074,54 @@ pub struct BuildVariant {
...
+    /// bzImage protocol, matching every `nitro-cli`-built EIF. Kept
+    /// selectable (rather than a hardcoded constant) because Firecracker
+    /// upstream has flipped between compressed and uncompressed kernels
+    /// more than once, and a bare-metal PVH loader still needs the ELF
+    /// `vmlinux` form.
+    ///
+    /// Ignored on aarch64 (only PE-wrapped `Image` is valid there) and on
+    /// non-EIF variants. Forwarded to `rpm2eif` as the
+    /// `EIF_KERNEL_FORMAT` build-arg (empty string when unset; `rpm2eif`
+    /// applies its own default).
```
Comment (ginglis13) [inline tools/buildsys/src/manifest.rs:1122]:
```suggestion
```

File: tools/buildsys/src/manifest.rs [hunk 26]
```diff
@@ -984,6 +1142,35 @@ pub enum ImageFormat {
     Vmdk,
 }
 
+/// x86_64 kernel image format that `rpm2eif` embeds in the EIF kernel
+/// section. See [`BuildVariant::eif_kernel_format`] for the rationale on
+/// keeping this selectable rather than pinning one value.
+///
+/// The TOML values `"bzimage"` and `"vmlinux"` are accepted; both are
+/// case-insensitive (via the `lowercase` rename). Ignored on aarch64.
```
Comment (ginglis13) [inline tools/buildsys/src/manifest.rs:1150]:
```suggestion
/// x86_64 kernel image format that `rpm2eif` embeds in the EIF kernel
/// section.
```

File: tools/buildsys/src/manifest.rs [hunk 27]
```diff
@@ -984,6 +1142,35 @@ pub enum ImageFormat {
...
+#[derive(Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+pub enum EifKernelFormat {
+    /// Compressed `vmlinuz` embedded verbatim. Matches the boot protocol
+    /// the sidecar Nitro Enclaves in-memory image loader expects (and
+    /// every `nitro-cli`-built EIF).
```
Comment (ginglis13) [inline tools/buildsys/src/manifest.rs:1156]:
```suggestion
    /// Compressed `vmlinuz` embedded verbatim.
```

File: tools/eif-builder/src/lib.rs [hunk 28]
```diff
@@ -7,24 +7,49 @@
...
 const EIF_SECTION_KERNEL: u16 = 1;
 const EIF_SECTION_CMDLINE: u16 = 2;
 const EIF_SECTION_RAMDISK: u16 = 3;
+const EIF_SECTION_SIGNATURE: u16 = 4;
 const EIF_SECTION_METADATA: u16 = 5;
 
+/// Human-readable name for an EIF section type. `"unknown"` for values that
+/// aren't one of the five defined section types; the caller can decide how
+/// to surface those (we don't reject them, so `describe_eif` can still report
+/// what's on disk for post-mortem debugging of a malformed file).
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:41]:
```suggestion
/// Human-readable name for an EIF section type. `"unknown"` for values that
/// aren't one of the five defined section types
```

File: tools/eif-builder/src/lib.rs [hunk 29]
```diff
@@ -33,11 +58,39 @@ const EIF_HEADER_SIZE: usize = 548;
...
+/// or, preferably, via `[package.metadata.build-variant] eif-pcie-flags` in
+/// the variant's `Cargo.toml`, which the buildsys → `rpm2eif` / `eif2eif`
+/// pipeline forwards to `--pcie-flags` on both first-build and repack. See
+/// `tools/buildsys/src/manifest.rs::BuildVariant::eif_pcie_flags`.
+///
+/// Longer-term plan: have the shim derive its launch flags from the header
+/// directly (host echoes, hypervisor enforces). That removes the coupling
+/// and this default becomes truly authoritative rather than
+/// coincidentally-agreed. The `--pcie-flags` knob is still needed as the
+/// per-variant authoring surface for the header value itself.
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:93]: I don't think this is the right place for this comment, esp discussions of "longer term plan". Maybe in `docs/design`.

```suggestion
```

File: tools/eif-builder/src/lib.rs [hunk 30]
```diff
@@ -286,26 +352,731 @@ pub fn build_eif(
...
+
+/// Summary of an EIF's contents as a `serde_json::Value`.
+///
+/// Consumers (notably `eif2eif`) use this to carry `BuildMetadata` forward
+/// across a repack: rpm2eif populates `KernelVersion` and `BuildToolVersion`
+/// from the source rootfs' installed RPMs; eif2eif has no rootfs to query,
+/// so it reads them out of the input EIF via this function and forwards them
+/// on the `eif-builder build` invocation that produces the new EIF.
+///
+/// Output shape (stable — this JSON is a public contract):
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:694]:
```suggestion
/// Output shape:
```

File: tools/eif-builder/src/lib.rs [hunk 31]
```diff
@@ -286,26 +352,731 @@ pub fn build_eif(
...
+///       ...
+///     },
+///     ...
+///   }
+/// }
+/// ```
+///
+/// `metadata` is `null` when the EIF has no METADATA section (rare — every
+/// EIF our pipeline produces has one, but a hand-crafted or foreign-tool EIF
+/// might not). `cmdline` is the CMDLINE section decoded as UTF-8 with a
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:723]:
```suggestion
/// `metadata` is `null` when the EIF has no METADATA section. `cmdline` is the CMDLINE section decoded as UTF-8 with a
```

File: tools/eif-builder/src/lib.rs [hunk 32]
```diff
@@ -286,26 +352,731 @@ pub fn build_eif(
...
+///
+/// Errors:
+/// - `ReadInput`  — I/O reading `input`.
+/// - `ParseInput` — malformed header/section table (via [`read_sections`]).
+/// - `MetadataDecode` — METADATA section is not valid UTF-8 JSON.
+///
+/// Note: CMDLINE is not required to be UTF-8, but Bottlerocket's rpm2eif
+/// always writes it as ASCII/UTF-8. If a non-UTF-8 CMDLINE ever appears here,
+/// we fall back to the lossy replacement encoding rather than fail — the
+/// primary consumer (metadata carry-forward) does not depend on CMDLINE.
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:735]:
```suggestion
```

File: tools/eif-builder/src/lib.rs [hunk 33]
```diff
@@ -286,26 +352,731 @@ pub fn build_eif(
...
+/// ```
+///
+/// Because the comparison is byte-exact, our payload must serialize to the
+/// *same* CBOR bytes as upstream's `PcrInfo`. `serde_cbor` encodes `Vec<u8>`
+/// as a CBOR **array of integers** (major type 4), not a byte string — the
+/// standard serde `Vec<T>` path calls `serialize_seq`. That is the wire
+/// shape upstream produces, and the shape this struct produces here.
+///
+/// Field names and order must match upstream exactly. `register_index` is
+/// `i32` upstream — do not widen.
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:940]:
```suggestion
```

File: tools/eif-builder/src/lib.rs [hunk 34]
```diff
@@ -286,26 +352,731 @@ pub fn build_eif(
...
+}
+
+/// Local mirror of upstream `aws-nitro-enclaves-image-format::defs::PcrSignature`.
+///
+/// The outer envelope of the `EifSectionSignature` is a `Vec<PcrSignature>`
+/// serialized with `serde_cbor::to_vec`. The upstream `describe-eif` path
+/// decodes it with `serde_cbor::from_slice::<Vec<PcrSignature>>`, so we
+/// must serialize with the *same* serde derive on a struct with the *same*
+/// field shape (`Vec<u8>` — not `serde_bytes::ByteBuf`), or the decode will
+/// fail with `invalid type: byte array, expected a sequence`.
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:954]:
```suggestion
```

File: tools/eif-builder/src/lib.rs [hunk 35]
```diff
@@ -286,26 +352,731 @@ pub fn build_eif(
...
+///
+/// Note: the byte fields decode as CBOR **arrays of integers** (major
+/// type 4), not byte strings (major type 2). This is what `serde_cbor`
+/// emits for a `Vec<u8>` under a plain `#[derive(Serialize)]` — the
+/// standard serde `Vec<T>` path — and it is what upstream produces and
+/// upstream consumers expect. Emitting byte strings here (via
+/// `ciborium::Value::Bytes` or `serde_bytes`) would make the CBOR envelope
+/// unparseable by `serde_cbor::from_slice::<Vec<PcrSignature>>` and would
+/// desync the COSE payload from the verifier's `measured_payload`
+/// recomputation, silently failing signature verification.
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:983]:
```suggestion
```

File: tools/eif-builder/src/lib.rs [hunk 36]
```diff
@@ -286,26 +352,731 @@ pub fn build_eif(
...
+
+/// Shared DER helpers for test code. `wrap_der` produces a DER TLV with
+/// short/long-form length encoding; used by both `signer.rs` (for SPKI
+/// synthesis) and by `lib.rs` tests (for the minimal self-signed cert).
+///
+/// Kept in one place so the length-encoding logic can't drift across the
+/// crate — the family of "hand-rolled DER" was flagged in code review as a
+/// maintenance risk when duplicated.
+#[cfg(test)]
+pub(crate) mod der_helpers {
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:1056]: why not just define these directly in `mod tests` ?

File: tools/eif-builder/src/lib.rs [hunk 37]
```diff
@@ -286,26 +352,731 @@ pub fn build_eif(
...
+    Ok(out)
+}
+
+/// Shared DER helpers for test code. `wrap_der` produces a DER TLV with
+/// short/long-form length encoding; used by both `signer.rs` (for SPKI
+/// synthesis) and by `lib.rs` tests (for the minimal self-signed cert).
+///
+/// Kept in one place so the length-encoding logic can't drift across the
+/// crate — the family of "hand-rolled DER" was flagged in code review as a
+/// maintenance risk when duplicated.
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:1054]:
```suggestion
```

File: tools/eif-builder/src/main.rs [hunk 38]
```diff
@@ -104,23 +215,95 @@ fn write_prepared_kernel(
...
-        args.cpus,
-        args.pcie_flags,
-        args.arch,
-        &metadata,
-    )?;
-    println!("Created {}", args.output.display());
+
+    // Dispatch: three states — unsigned, local-signed, KMS-signed. Clap has
+    // already enforced the mutual-exclusion constraints, so we only need to
+    // build the right signer here.
```
Comment (ginglis13) [inline tools/eif-builder/src/main.rs:247]:
```suggestion
```

File: tools/eif-builder/src/lib.rs [hunk 39]
```diff
@@ -497,4 +1268,906 @@ mod tests {
...
+        assert_eq!(envelope.len(), 1);
+        let cose_bytes = &envelope[0].signature;
+
+        // Extract the COSE payload...
+        let sign1 = CoseSign1::from_slice(cose_bytes).unwrap();
+        let coses_payload = sign1.payload.expect("payload must be present");
+
+        // ...recompute PCR0 the same way `build_signed_eif` did, then
+        // reproduce the upstream `measured_payload` construction exactly.
+        let pcr0 = compute_pcr0(b"FAKE_KERNEL", b"cmd", &[&[][..]]);
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs:1622]: Can we use the upstream pcr measurement logic? https://github.com/aws/aws-nitro-enclaves-image-format/blob/main/src/utils/mod.rs

**Comment by jmt-lab** [comment]
We have recreated this review as a GH PR stack finalizing here: https://github.com/bottlerocket-os/twoliter/pull/722
Please use the branch for that pr (the last in the stack for testing).
